### PR TITLE
feat(testing): OHIF Test Agent Skills

### PR DESCRIPTION
**Pull request by diattamo**
## Summary

Adds an `ohif-test-agent` skill that teaches AI coding agents how to generate correct, runnable Playwright E2E tests for the OHIF Viewer using its custom fixture system, page objects, normalized WebGL viewport coordinates, and visual-regression conventions.

This PR landed iteratively. The original proposal published the skill under `tests/ohif-test-agent/` and required each contributor to manually copy or symlink the folder into their agent client's skill directory. **The current shape replaces that with a single canonical location that all major clients discover automatically**, plus reviewer-driven hardening of the rules the skill enforces.

## Changes

The skill is at `.agents/skills/ohif-test-agent/` and contains:

- **`SKILL.md`** — the entire behavior contract. Covers the environment model, test-authoring workflow, architecture (custom fixtures, page objects, normalized coordinates, WebGL viewport), study loading lifecycle, render-cycle waits vs sleeps, visual regression rules, Playwright config facts, canonical study UIDs, a hard rules list, and a mandatory pre-output self-check.
- **`assets/spec-template.ts`** — a ready-to-copy test template with the correct imports, fixture destructuring, viewport interaction patterns, prompt handling, and assertion strategies.
- **`references/patterns-by-feature.md`** — feature-area-indexed seed specs (length / freehand / SEG hydration / RTSTRUCT / TMTV / MPR / crosshairs / overlays / tag browser / etc.) plus the canonical study UID table.
- **`references/page-objects.md`** — stable structural rules for the page object system; deliberately does not enumerate methods (the source is authoritative).
- **`references/utilities.md`** — stable rules for `tests/utils/` (barrel vs deep imports, the `@playwright/test` trap, object-param convention, `visitStudy` delays, `checkForScreenshot` direction).
- **`references/failure-triage.md`** — categorize-then-debug guide for the failures the suite actually produces.

## Discovery by agents: zero-config for three clients, one-liner for Claude Code

| Client | Discovery | Setup required |
|---|---|---|
| GitHub Copilot / VS Code Agent | Scans `.agents/skills/` natively | none |
| OpenAI Codex (CLI / IDE per docs) | Scans `.agents/skills/` natively | none |
| Cursor | Scans `.agents/skills/` natively (per [Cursor docs](https://cursor.com/docs/skills.md)) | none |
| Claude Code | Scans only `.claude/skills/` | one-line symlink, documented in `AGENTS.md` |

A root `AGENTS.md` was added in another commit. This has been updated to include a short "Skills" section describing the test skill and the Claude Code setup command (Windows variant included).

## Working examples:
| Client | Screenshot |
|---|---|
| gh Copilot | image |
| Cursor| image |
| Claude | image |
| Codex | `/<skill>` works after putting the skill files in a .cursor director similar to .claude. However, unlike claude, in the current .agents directory, it is discovered and used during agent task performance|
| (even) Antigravity | image |

## Test plan

- [ ] Pull branch
- [ ] Confirm `.agents/skills/ohif-test-agent/SKILL.md` exists
- [ ] In **VS Code with Copilot Agent**: type `/skills` in chat and confirm `ohif-test-agent` is listed
- [ ] In **Cursor**: type `/skills` in chat and confirm `ohif-test-agent` is listed
- [ ] In **Claude Code**: run the documented `ln -s` once, then start a new session and confirm `ohif-test-agent` is in the discovered-skills list
- [ ] Prompt: *"Write an OHIF Playwright test for the Length tool"* — verify the generated test
- [ ] On Windows (if available): confirm `core.symlinks=true` allows the `.claude/skills/ohif-test-agent` symlink to materialize correctly

## Summary by CodeRabbit

* **Documentation**
  * Added the ohif-test-agent behavior contract, including the required test workflow, output expectations, and stable WebGL/render-interaction guidance.
  * Introduced a standardized Playwright OHIF visual/DOM regression spec template covering study lifecycle, viewport render synchronization, and assertion patterns.
  * Added/updated reference guides for page objects, utilities, feature-area seed specs (including an options/hotkeys area), and a structured failure triage guide.
  * Documented the skill location in AGENTS.md.
* **Chores**
  * Updated preinstall to mirror skill docs into local tooling via symlinks.
  * Updated .gitignore to ignore the local tooling directory and moved the rule earlier.

<h3>Greptile Summary</h3>

This PR introduces the `ohif-test-agent` skill — a structured knowledge package that teaches AI coding agents how to generate correct, runnable Playwright E2E tests for the OHIF Viewer. It adds auto-discovery support for GitHub Copilot, Cursor, and OpenAI Codex natively via `.agents/skills/`, and automates the Claude Code symlink via `preinstall.js`.

- **`SKILL.md` + reference files** (`page-objects.md`, `utilities.md`, `patterns-by-feature.md`, `failure-triage.md`): full behavior contract covering the custom fixture system, normalized WebGL coordinates, render-cycle wait idioms, hydration prompt handling, and visual regression rules — all internally consistent and cross-referenced.
- **`assets/spec-template.ts`**: ready-to-copy scaffold using the object-form `checkForScreenshot`, correct barrel imports, and the `waitForViewportRenderCycle` pattern.
- **`preinstall.js`**: mirrors `.agents/skills/*/` into `.claude/skills/*/` via relative-path symlinks with idempotency checking; on Windows without symlink privileges `fs.symlinkSync` throws EPERM (caught and logged), so the manual `ln -s` fallback documented in AGENTS.md still applies on restricted Windows systems.

<h3>Confidence Score: 5/5</h3>

All changes are documentation and developer-tooling; no production code paths are touched.

The skill files are internally consistent, the spec template uses the correct object-form patterns, the `preinstall.js` mirroring logic handles all edge cases gracefully with try-catch, and no application or test runtime code is modified.

No files require special attention; previously flagged issues in `.gitignore` and the earlier draft of `spec-template.ts` appear addressed in the current HEAD.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| .agents/skills/ohif-test-agent/SKILL.md | Adds the full OHIF test-agent behavior contract: environment model, fixture architecture, normalized viewport interactions, render-cycle wait idioms, visual regression rules, study lifecycle, and a mandatory pre-output self-check. Content is internally consistent and aligns with the reference files. |
| .agents/skills/ohif-test-agent/assets/spec-template.ts | Ready-to-copy test scaffold using the correct object-form `checkForScreenshot`, fixture destructuring, render-cycle watcher idiom, and barrel imports. No legacy patterns present. |
| .agents/skills/ohif-test-agent/references/utilities.md | Documents barrel vs. deep import rules, the `@playwright/test` trap, `visitStudy` delay conventions, and the object-form `checkForScreenshot` mandate. The `screenShotPaths` usage example now correctly uses the object form with a `locator`. |
| preinstall.js | Adds skill mirroring from `.agents/skills/` to `.claude/skills/` via relative-path symlinks. Idempotency check compares stored symlink target against expected; graceful error handling throughout. On Windows without symlink privileges, `fs.symlinkSync` throws EPERM (caught and logged) rather than silently falling back. |
| .gitignore | Removes `**/.claude/settings.local.json`; subpackage `.claude/settings.local.json` files are no longer gitignored (previously flagged in review thread). |
| AGENTS.md | Adds a four-line Skills section pointing agents to the skill location — minimal and correct. |

</details>

</details>

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `.agents/skills/ohif-test-agent/references/utilities.md`, line 722-724 ([link](https://github.com/ohif/viewers/blob/f87a492e246915a2eb8b671a9a45881bf6d7dd7f/.agents/skills/ohif-test-agent/references/utilities.md#L722-L724)) 

   <a href="#">P2</a> **`screenShotPaths` usage example contradicts the object-form mandate just above it**

   The example in this section shows `checkForScreenshot(page, page, screenShotPaths.length.lengthDisplayedCorrectly)` — the positional form — immediately after the file mandates the object form for all new specs and tells agents not to copy the positional pattern from existing code. An agent scanning for "how do I use `screenShotPaths`?" will land on this line and reproduce the legacy call. The example should match the required object form with a `normalizedClip`.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: .agents/skills/ohif-test-agent/references/utilities.md
   Line: 722-724

   Comment:
   **`screenShotPaths` usage example contradicts the object-form mandate just above it**

   The example in this section shows `checkForScreenshot(page, page, screenShotPaths.length.lengthDisplayedCorrectly)` — the positional form — immediately after the file mandates the object form for all new specs and tells agents not to copy the positional pattern from existing code. An agent scanning for "how do I use `screenShotPaths`?" will land on this line and reproduce the legacy call. The example should match the required object form with a `normalizedClip`.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<sub>Reviews (5): Last reviewed commit: ["docs(test-agent): generalize uncovered-f..."](https://github.com/ohif/viewers/commit/30eeddff46314d0b07a0804a1c239d93df8cdd42) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34633970)</sub>

### CODE REVIEW

File: tests/ohif-test-agent/instructions.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+5. Handle prompt flows when relevant
+6. Include static-check notes if execution was not performed
+
+---
+
+## Claude setup (starting point)
+
+1. Locate your Claude skills directory:
+   - Common location: `~/.claude/skills/`
+2. Copy this folder into that directory:
```
Comment (jbocce) [inline tests/ohif-test-agent/instructions.md]: Is there any way for Claude to just find it in the worktree's directory? I apologize if this is a stupid question.

File: tests/ohif-test-agent/instructions.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Notes:
+
+- Claude triggering uses the `name` and `description` frontmatter in `SKILL.md`.
+- Keep `references/` and `assets/` alongside `SKILL.md`; links assume this structure.
+
+---
+
+## Codex setup
+
+Point your Codex runtime at `.github/agents/ohif-test-agent/`.
```
Comment (jbocce) [inline tests/ohif-test-agent/instructions.md]: Do we need instructions prior to this for copying ohif-test-agent to the .github directory?

File: tests/ohif-test-agent/instructions.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+5. Test with a prompt that clearly targets OHIF Playwright E2E behavior.
+
+Notes:
+
+- Claude triggering uses the `name` and `description` frontmatter in `SKILL.md`.
+- Keep `references/` and `assets/` alongside `SKILL.md`; links assume this structure.
+
+---
+
+## Codex setup
```
Comment (jbocce) [inline tests/ohif-test-agent/instructions.md]: Can we add something for Cursor too?

File: .agents/skills/ohif-test-agent/references/patterns-by-feature.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+## Canonical study UIDs
+
+| UID | Mode(s) | What it has |
+|-----|---------|-------------|
+| `1.3.6.1.4.1.25403.345050719074.3824.20170125095438.5` | `viewer` | CT — default for measurement/annotation tests |
+| `1.3.6.1.4.1.14519.5.2.1.1706.8374.643249677828306008300337414785` | `viewer` | CT volume for 3D/MPR/crosshairs |
+| `1.3.6.1.4.1.14519.5.2.1.256467663913010332776401703474716742458` | `viewer` or `segmentation` | CT + SEG for labelmap tests |
+| `1.2.840.113619.2.290.3.3767434740.226.1600859119.501` | `viewer` / `segmentation` / `tmtv` | CT + RTSTRUCT + PET, used for contour and TMTV |
+| `1.3.6.1.4.1.14519.5.2.1.7695.4007.324475281161490036195179843543` | `viewer` | SR structured report |
```
Comment (jbocce) [inline .agents/skills/ohif-test-agent/references/patterns-by-feature.md:17]: We should consider adding more here.

File: .agents/skills/ohif-test-agent/references/patterns-by-feature.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+Do not invent UIDs — they must exist on the e2e data server.
```
Comment (jbocce) [inline .agents/skills/ohif-test-agent/references/patterns-by-feature.md:19]: Good call.

File: tests/ohif-test-agent/references/utilities.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+So: pick `10000` when the mode is `tmtv`, `2000` otherwise, and only ramp up if a specific test flakes on first-render assertions. The delay is a good first lever for "not visible" flakes, but it's not a universal upgrade.
+
+### `checkForScreenshot` has two call forms
+
+- **Object form** (preferred in newer specs): `checkForScreenshot({ page, screenshotPath, maxDiffPixelRatio?, threshold?, normalizedClip?, fullPage? })`
+- **Positional form** (still in use across older specs): `checkForScreenshot(page, locator, screenshotPath, ...)`
+
+The object form exposes tuning knobs the positional form doesn't — `maxDiffPixelRatio`, `threshold`, `normalizedClip`. For 3D content, bump `maxDiffPixelRatio` to around `0.04` because GPU rendering is noisier. Check the current signature in [tests/utils/checkForScreenshot.ts](tests/utils/checkForScreenshot.ts) (or wherever the barrel points) if something looks off.
```
Comment (jbocce) [inline tests/ohif-test-agent/references/utilities.md]: I would rather AI not monkey with maxDiffPixelRatio. That should be something rarely touched.

File: tests/ohif-test-agent/references/utilities.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+The function's own default delay is `0`. Nearly every spec passes `2000` to let the first render settle. The one consistent exception is `mode: 'tmtv'`, where the suite uniformly uses `10000` because PET fusion and SUV calculation add real wall-clock cost before the UI is interactive.
+
+Notably, 3D layouts in `viewer` mode (3DOnly, 3DFourUp, 3DMain, 3DPrimary) and MPR also use `2000` — they're not "heavy" in the `visitStudy` sense. Their stabilization problem is solved at the interaction layer with `attemptAction(() => reduce3DViewportSize(page), 10, 100)`, not by a longer visit delay.
+
+So: pick `10000` when the mode is `tmtv`, `2000` otherwise, and only ramp up if a specific test flakes on first-render assertions. The delay is a good first lever for "not visible" flakes, but it's not a universal upgrade.
+
+### `checkForScreenshot` has two call forms
+
+- **Object form** (preferred in newer specs): `checkForScreenshot({ page, screenshotPath, maxDiffPixelRatio?, threshold?, normalizedClip?, fullPage? })`
```
Comment (jbocce) [inline tests/ohif-test-agent/references/utilities.md]: The only reason object form is preferred because we want to stop taking the screenshot of the entire app and instead focus on the viewport or viewport grid. This should somehow be worked in here.

File: tests/ohif-test-agent/instructions.md (hunk 1 above)
Comment (diattamo) [reply tests/ohif-test-agent/instructions.md]: I updated the overall configuration to not have to perform this anymore. 
But to answer your question. I have now made it so that claude finds it in the worktree. Just need to run a simple one line command.

File: tests/ohif-test-agent/instructions.md (hunk 2 above)
Comment (diattamo) [reply tests/ohif-test-agent/instructions.md]: Not needed anymore

File: tests/ohif-test-agent/instructions.md (hunk 3 above)
Comment (diattamo) [reply tests/ohif-test-agent/instructions.md]: Done. Cursor should now work out of the box.

File: tests/ohif-test-agent/references/utilities.md (hunk 6 above)
Comment (diattamo) [reply tests/ohif-test-agent/references/utilities.md]: Updated the instruction to not mess with this. I added instruction to have it reviewed instead.

File: tests/ohif-test-agent/references/utilities.md (hunk 7 above)
Comment (diattamo) [reply tests/ohif-test-agent/references/utilities.md]: Updated this basically default to not using screenshots going forward. Take a look at the wording.

File: AGENTS.md [hunk 8]
```diff
@@ -176,6 +176,18 @@ Always prioritrize pub sub, by calling a services subscribe over useEffects as i
 ### Never modify core architecture
 Do not modify the core and always find a way to implement the solution via the extensions and modes, only modify core as a last resort if all other fail or there's an architectural constraint.
 
+## Skills
+
+The `ohif-test-agent` skill (Playwright E2E test guidance) lives at `.agents/skills/ohif-test-agent/`. GitHub Copilot/VS Code Agent, OpenAI Codex, and Cursor scan `.agents/skills/` natively — no setup required. Claude Code scans only `.claude/skills/` so set up the symlink locally once after cloning:
```
Comment (jbocce) [inline AGENTS.md]: Can this be tied into the preinstall.js file like what is done for (this) existing AGENTS.md file? I believe there is already a (similar) auto link for the AGENTS.md file in there.

Comment (diattamo) [reply AGENTS.md]: Implemented so that it should work with
`yarn install (or preinstall)`

**Comment by jbocce** [comment]
Ok I tried this branch with Claude and had it write a test for pan. It looks pretty good.

The only thing I did not understand is why it added the `normalizedClip` parameter. There is no need for that.

I am next going to try having it write a test where some PageObjects either need to be created or augmented.
```
import {
  checkForScreenshot,
  screenShotPaths,
  test,
  visitStudy,
  waitForViewportRenderCycle,
} from './utils';

test.beforeEach(async ({ page }) => {
  const studyInstanceUID = '1.3.6.1.4.1.25403.345050719074.3824.20170125095438.5';
  const mode = 'viewer';
  await visitStudy(page, studyInstanceUID, mode, 2000);
});

test('should pan the image when dragging with the Pan tool active', async ({
  page,
  mainToolbarPageObject,
  viewportPageObject,
}) => {
  await mainToolbarPageObject.panTool.click();

  const activeViewport = await viewportPageObject.active;

  const renderCycle = waitForViewportRenderCycle(page);
  await activeViewport.normalizedDragAt({
    start: { x: 0.5, y: 0.5 },
    end: { x: 0.7, y: 0.7 },
  });
  await renderCycle;

  await checkForScreenshot({
    page,
    locator: activeViewport.pane,
    screenshotPath: screenShotPaths.pan.panDisplayedCorrectly,
    normalizedClip: { x: 0, y: 0, width: 1, height: 1 },
  });
});
```

**Comment by jbocce** [comment]
I also asked it to "Write a test for setting the zoom user preference hot key."

It was good to prompt me to add `data-cy` attributes, however the test it wrote was "ok". The test itself would NOT pass the muster for several reasons...

1. It did not create page object or page object methods for...
     1. the options menu and its menu items
     2. the preferences dialog which probably should have been added to the `DOMOverlayPageObject`
     3. each preference in the dialog should be some type of entry in the page object created for the preference dialogue
     4. zoom button should have been added to the `MainToolbarPageObject`
2. besides checking the zoom button to have the `data-active` attribute, it probably should have dragged on the viewport to check if the image zoomed or not

I would say it did as well as an entry level programmer that did not check how other tests are written.

Here is the test...

```
import { expect, test, visitStudy } from './utils';

test.beforeEach(async ({ page }) => {
  const studyInstanceUID = '1.3.6.1.4.1.25403.345050719074.3824.20170125095438.5';
  const mode = 'viewer';
  await visitStudy(page, studyInstanceUID, mode, 2000);
});

test('should activate the Zoom tool via a user-defined hotkey set in User Preferences', async ({
  page,
}) => {
  await page.getByTestId('options-menu').click();
  await page.getByTestId('menu-item-settings').click();

  const modal = page.getByRole('dialog');
  await expect(modal).toBeVisible();

  const zoomHotkeyInput = page.getByTestId('hotkey-input-Zoom');
  await zoomHotkeyInput.focus();
  await page.keyboard.press('q');
  await expect(zoomHotkeyInput).toHaveValue('Q');

  await page.getByRole('button', { name: 'Save' }).click();
  await expect(modal).toBeHidden();

  await page.keyboard.press('q');
  await expect(page.getByTestId('Zoom')).toHaveAttribute('data-active', 'true');
});
```

**Review by jbocce — CHANGES_REQUESTED** [review]
So I think Claude did ok with minimal instructions, but either the various `*.md` files need more info or simply Claude needs to be given more instructions. Either way I will let you update the PR as you see fit and then we can think about merging this.

**Comment by jbocce** [comment]
@sedghi would like to have a look at this before we merge? Do you have any concerns with this? Thanks.

File: AGENTS.md [hunk 9]
```diff
@@ -176,6 +176,12 @@ Always prioritrize pub sub, by calling a services subscribe over useEffects as i
 ### Never modify core architecture
 Do not modify the core and always find a way to implement the solution via the extensions and modes, only modify core as a last resort if all other fail or there's an architectural constraint.
 
+## Skills
+
+The `ohif-test-agent` skill (Playwright E2E test guidance) lives at `.agents/skills/ohif-test-agent/`. GitHub Copilot/VS Code Agent, OpenAI Codex, and Cursor scan `.agents/skills/` natively. Since Claude Code scans only `.claude/skills/`, `preinstall.js` automatically symlinks `.agents/skills/` into `.claude/skills/` during `yarn install`. The symlinks point at the canonical skills, so future updates are picked up automatically and new skills added under `.agents/skills/` are linked on the next install.
```
Comment (sedghi) [inline AGENTS.md]: This reads like a readme than agents instruction
I would just remove any extra explanation (where github looks etc), just say where is the skill and that should be enough for the agent

File: .agents/skills/ohif-test-agent/assets/spec-template.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    //    instead of sleeping — start the watcher BEFORE the action:
+    //    const cycle = waitForViewportRenderCycle(page);
+    //    await action();
+    //    await cycle;
+
+    // 3. Handle prompts (first measurement prompts for tracking; SEG/RT/SR prompts for hydration)
+    //    await DOMOverlayPageObject.viewport.measurementTracking.confirm.click();
+
+    // 4. Assert canvas output via visual regression
+    //    await checkForScreenshot(page, page, screenShotPaths.YOUR_CATEGORY.YOUR_KEY);
```
Comment (sedghi) [inline .agents/skills/ohif-test-agent/assets/spec-template.ts]: please check coderabbit, i think it has something useful commented here. no?

Comment (diattamo) [reply .agents/skills/ohif-test-agent/assets/spec-template.ts]: Resolved! the method format was indeed not accurate. Updated!

File: AGENTS.md (hunk 9 above)
Comment (diattamo) [reply AGENTS.md]: I removed the verbose instructions.

**Review by sedghi — APPROVED** [review]

File: .agents/skills/ohif-test-agent/references/patterns-by-feature.md [hunk 11]
```diff
@@ -114,6 +114,21 @@ These are the rare specs that use `page.goto(...)` directly instead of `visitStu
 
 Same shape as SEG hydration but load via `loadSeriesByModality('SR')`.
 
+## 16. User preferences / hotkeys / options menu / settings dialog
```
Comment (jbocce) [inline .agents/skills/ohif-test-agent/references/patterns-by-feature.md]: Wait... does this mean we need a section for each area that is not covered yet? If so, that sucks.

Comment (diattamo) [reply .agents/skills/ohif-test-agent/references/patterns-by-feature.md]: I updated it to make it explicit that the idea is not to add more sections. It documents the existing ones for the agents to load into context. and provide explicit examples to follow

**Review by jbocce — APPROVED** [review]
